### PR TITLE
[c++] Distance-agnostic triplet margin loss

### LINKED ISSUE #43342: Proposal: Generic Triplet-Margin Loss
## 🚀 Feature
We'd like to add a distance-agnostic version of the [TripletMarginLoss](https://pytorch.org/docs/master/generated/torch.nn.TripletMarginLoss.html), as the current version only supports l_p norms.  This was something we implemented as part of a recent investigation for embedding learning and thought could be helpful to the PyTorch userbase at large.

## Motivation

1. Extensibility: decoupling the distance function and the loss computation allow for more flexibility for users.  PyTorch currently has a [CosineEmbeddingLoss](https://pytorch.org/docs/stable/generated/torch.nn.CosineEmbeddingLoss.html), but that serves a somewhat different purpose and doesn't really work for users wanting a triplet-margin loss with cosine distance.

2. Existing use cases: several papers have proposed triplet loss functions with cosine distance ([1](https://www.isca-speech.org/archive/Interspeech_2018/pdfs/1209.pdf), [2](https://www.researchgate.net/publication/316736728_Deep_Speaker_an_End-to-End_Neural_Speaker_Embedding_System#read)) or have generally used cosine-based metrics ([1](https://openaccess.thecvf.com/content_cvpr_2018/papers/Wang_CosFace_Large_Margin_CVPR_2018_paper.pdf), [2](https://arxiv.org/pdf/1812.00442.pdf)).  [PyTorch-BigGraph](https://arxiv.org/abs/1903.12287) also does something similar with its ranking loss.

3.  Products like Tensorflow [decouple](https://github.com/tensorflow/addons/blob/v0.10.0/tensorflow_addons/losses/triplet.py) the distance functions and even allow for custom distance metrics

## Pitch

We have a working implementation of this (see screenshots) and propose to use it directly, with adjustments to fit the existing PyTorch code structure.  We'll tentatively call these `torch.nn.GenericTripletMarginLoss` and `F.generic_triplet_margin_loss`.  In the proposed implementation, `GenericTripletMarginLoss` will be a stateful wrapper for `F.generic_triplet_margin_loss`, and the loss computation will occur in the latter.

The module signatures will resemble that of the [production module](https://pytorch.org/docs/master/generated/torch.nn.TripletMarginLoss.html), with an analogous signature for the [functional module](https://pytorch.org/docs/master/_modules/torch/nn/modules/loss.html#TripletMarginLoss).  The removed parameters (`p` and `eps`) are specific to l_p norm computation; `size_average` and `reduce` are removed because of deprecation.  There are two new parameters: `distance_function` and `is_similarity_function`.  `distance_function` is a function/module $$f: R^{n \cross d} \rightarrow R^n$$ that gives an index-wise distance metric between two tensors of embeddings.  This can be, for example, `torch.nn.PairwiseDistance` or `torch.nn.CosineSimilarity`.  `is_similarity_function` is a boolean that denotes whether `distance_function` is a distance or similarity function: this will indicate whether to flip the signs in distance computation.

## Alternatives

Beyond the straightforward alternative of leaving this to the user to implement every time, we haven't come up with any other alternatives.

One open question is whether the Python implementation is performant enough, or if this would need to be added in aTen as well.

## Additional context

image

cc @albanD @mruberry

### PR DESCRIPTION
**Pull request by ethch18**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#45377 [c++] Distance-agnostic triplet margin loss**

This PR adds a C++ implementation of the TripletMarginWithDistanceLoss, for which the Python implementation was introduced in PR #43680.  It's based on PR #44072, but I'm resubmitting this to unlink it from Phabricator.

cc: @mruberry @albanD @glaringlee

Differential Revision: [D24003973](https://our.internmc.facebook.com/intern/diff/D24003973)

### CODE REVIEW

**Comment by glaringlee** [comment]
@ethch18 
I'll rebase this soon. Can you do a rebase? It seems the CI tests failed due to a problematic master base.

File: torch/csrc/api/src/nn/modules/loss.cpp [hunk 1]
```diff
@@ -180,6 +180,33 @@ Tensor TripletMarginLossImpl::forward(
 
 // ============================================================================
 
+TripletMarginWithDistanceLossImpl::TripletMarginWithDistanceLossImpl(
+    const TripletMarginWithDistanceLossOptions& options_) // NOLINT(modernize-pass-by-value)
```
Comment (glaringlee) [inline torch/csrc/api/src/nn/modules/loss.cpp:184]: Will complier complain if removing `NOLINT(...)`?
I kinda remember c10:optional doesn't work well with std::move. Did you try std::move here?

Comment (ethch18) [reply torch/csrc/api/src/nn/modules/loss.cpp:184]: clint complains about this otherwise; I copied the `NOLINT()` from elsewhere in the file.

What would `std::move` look like in this case?  Looks like everything else in this file uses the same format, minus `NOLINT()` if it's not applicable.

File: test/cpp/api/modules.cpp [hunk 2]
```diff
@@ -2085,6 +2085,109 @@ TEST_F(ModulesTest, TripletMarginLoss) {
...
+        auto distanceOptions =
+            TripletMarginWithDistanceLossOptions()
+                .reduction(reduction)
+                .margin(margin)
+                .swap(swap);
+        TripletMarginLoss basicLoss(basicOptions);
+        TripletMarginWithDistanceLoss distanceLoss(distanceOptions);
+
+        auto basicOutput = basicLoss->forward(anchor, positive, negative);
+        auto distanceOutput = distanceLoss->forward(anchor, positive, negative);
```
[the remark this answers is not in the document]
Comment (ethch18) [reply test/cpp/api/modules.cpp:2121]: Does `()` mean, e.g., `distanceLoss(anchor, positive, negative)`?

[the remark this answers is not in the document]
Comment (glaringlee) [reply test/cpp/api/modules.cpp:2121]: Ah, sry, I am not clear. Yes.  distanceLoss(anchor, positive, negative).

File: torch/csrc/api/src/nn/modules/loss.cpp (hunk 1 above)
Comment (glaringlee) [reply torch/csrc/api/src/nn/modules/loss.cpp:184]: You can do the following to avoid that lint error.
`
TripletMarginWithDistanceLossImpl::TripletMarginWithDistanceLossImpl(
    TripletMarginWithDistanceLossOptions options_) : 
    options(std::move(options_)) {}
`
But I kinda remember std::move is not working well with c10::optional which is inside options. But if you can try, that would be great.

Comment (ethch18) [reply torch/csrc/api/src/nn/modules/loss.cpp]: std::move couldn't build once for me locally, then it started building -- not sure if it's a flaky failure.  I've pushed this and if the CI fails, then I'll go back to pass-by-reference

**Comment by glaringlee** [comment]
@ethch18 
All CI builds passed (rocm test failure is fine). I am going to build this on my end locally, will approve this if it passed. Thanks a lot for contributing.

**Comment by ethch18** [comment]
@glaringlee great!  FYI, Mike asked me to incorporate the changes from PR #45378 into this one too, which I'll do shortly

**Comment by glaringlee** [comment]
@ethch18 sure, go ahead!

**Comment by ethch18** [comment]
@glaringlee updated -- let me know if you need anything else from my end!

**Comment by glaringlee** [comment]
@ethch18 
Thanks, can you please rebase your code once? There are several CI test failed, I think that's upstream problems, rebase should solve it.

**Comment by ethch18** [comment]
Just rebased on fbcode/warm

**Comment by glaringlee** [comment]
@ethch18 
Sry, can you rebase again, this code still failed one of our CI test: pytorch_xla_linux_bionic_py3_6_clang9_build.

**Comment by ethch18** [comment]
@glaringlee rebased with fbcode/warm again.  If this doesn't work, I'll try viable/strict next

**Comment by glaringlee** [comment]
@ethch18 
Thanks, it should work. There is a know issue for our XLA build, it can be solved by rebase.

**Review by glaringlee — APPROVED** [review]
LGTM now
